package/resource_monitor/resource_monitor/src/query_process_file_descriptor.c

```diff
@@ -0,0 +1,491 @@
+  int rc = run_with_stdin_file(NULL, "sudo", argv, buf, sizeof(buf));
+  if (rc != 0) {
+#ifdef DEBUG_QUERY_FD_TAB
+    PK_LOG_ANNO(LOG_ERR,
+                "file descriptor query: error running 'lsof' command: %s",
+                strerror(errno));
+#endif
+    return;
+  }
+
+  char *line_ctx = NULL;
+
+  for (char *line = strtok_r(buf, "\n", &line_ctx); line != NULL;
+       line = strtok_r(NULL, "\n", &line_ctx)) {
+    if (!parse_fd_line(state, line)) {
+#ifdef DEBUG_QUERY_FD_TAB
+      PK_LOG_ANNO(LOG_ERR, "file descriptor query: parse_fd_line failed ");
+#endif
+      return;
+    }
+  }
+  size_t pid_count = table_count(state->pid_table);
+
+  id_to_count_t *NESTED_AXX(pid_to_count) = alloca(pid_count * sizeof(id_to_count_t));
+
+  table_foreach_key(state->pid_table,
+                    NULL,
+                    NESTED_FN(bool,
+                              (table_t * table, const char *key, size_t index, void *context1),
+                              {
+
+                                (void)context1;
+
+                                pid_entry_t *entry = pid_table_get(table, key);
+
+                                pid_to_count[index].id = key;
+                                pid_to_count[index].count = entry->fd_count;
+
+                                return true;
+                              }));
+
+  qsort(pid_to_count, pid_count, sizeof(id_to_count_t), compare_counts);
+
+  size_t file_count = table_count(state->file_table);
+
+  id_to_count_t *NESTED_AXX(file_to_count) = alloca(file_count * sizeof(id_to_count_t));
+
+  table_foreach_key(state->file_table,
+                    NULL,
+                    NESTED_FN(bool,
+                              (table_t * table, const char *key, size_t index, void *context1),
+                              {
+
+                                (void)context1;
+
+                                file_entry_t *entry = file_table_get(table, key);
+
+                                file_to_count[index].id = key;
+                                file_to_count[index].count = entry->fd_count;
+
+                                return true;
+                              }));
+
+  qsort(file_to_count, file_count, sizeof(id_to_count_t), compare_counts);
+
+
+  for (size_t i = 0; i < ITEM_COUNT && i < pid_count; i++) {
+    state->top10_pid_counts[i] = pid_to_count[i].id;
+#ifdef DEBUG_QUERY_FD_TAB
+    PK_LOG_ANNO(LOG_DEBUG, "TOP 10 FD PID : %s : %d", pid_to_count[i].id, pid_to_count[i].count);
+#endif
+  }
+  for (size_t i = 0; i < ITEM_COUNT && i < file_count; i++) {
+    state->top10_file_counts[i] = file_to_count[i].id;
+#ifdef DEBUG_QUERY_FD_TAB
+    PK_LOG_ANNO(LOG_DEBUG, "TOP 10 FD FILE : %s : %d", file_to_count[i].id, file_to_count[i].count);
+#endif
+  }
+}
+
+static bool query_fd_prepare(u16 *msg_type, u8 *len, u8 *sbp_buf, void *context)
+{
+  resq_state_t *state = context;
+
+  if (state->send_state == SEND_DONE) return false;
+
+  size_t index = (size_t)state->send_state;
```

Comment:
if we're going overload the values of `state->send_state` as both an enum and an index let's add the following assert:
```
/* Send states are re-used as index values */
assert(0 == SEND_FD_COUNTS_0);
```